test/util/test_app.go

```diff
@@ -73,23 +76,99 @@ func SetupTestAppWithGenesisValSet(cparams *tmproto.ConsensusParams, genAccounts
-	// var cache sdk.MultiStorePersistentCache
+// NewTestApp creates a new app instance with an empty memDB and a no-op logger.
+func NewTestApp() *app.App {
 	// EmptyAppOptions is a stub implementing AppOptions
 	emptyOpts := EmptyAppOptions{}
 	// var anteOpt = func(bapp *baseapp.BaseApp) { bapp.SetAnteHandler(nil) }
 	db := dbm.NewMemDB()
 
 	encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...)
 
-	testApp := app.New(
+	return app.New(
 		log.NewNopLogger(), db, nil,
 		cast.ToUint(emptyOpts.Get(server.FlagInvCheckPeriod)),
 		encCfg,
 		0,
 		emptyOpts,
 	)
+}
+
+// ApplyGenesisState sets genesis on initialized testApp with the provided arguments.
+func ApplyGenesisState(testApp *app.App, pubKeys []cryptotypes.PubKey, balance int64, cparams *tmproto.ConsensusParams) (keyring.Keyring, []genesis.Account, error) {
+	// create genesis
+	gen := genesis.NewDefaultGenesis().
+		WithChainID(ChainID).
+		WithConsensusParams(cparams).
+		WithGenesisTime(time.Date(2023, 1, 1, 1, 1, 1, 1, time.UTC).UTC())
+
+	// add accounts to genesis
+	for _, pk := range pubKeys {
+		err := gen.AddAccount(genesis.Account{
+			PubKey:  pk,
+			Balance: balance,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// add validator to genesis
+	err := AddDeterministicValidatorToGenesis(gen)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to add validator: %w", err)
+	}
+
+	genDoc, err := gen.Export()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to export genesis doc: %w", err)
+	}
+
+	// initialise test app against genesis
+	testApp.Info(abci.RequestInfo{})
+
+	abciParams := &abci.ConsensusParams{
+		Block: &abci.BlockParams{
+			// choose some value large enough to not bottleneck the max square
+			// size
+			MaxBytes: int64(appconsts.DefaultSquareSizeUpperBound*appconsts.DefaultSquareSizeUpperBound) * appconsts.ContinuationSparseShareContentSize,
+			MaxGas:   cparams.Block.MaxGas,
+		},
+		Evidence:  &cparams.Evidence,
+		Validator: &cparams.Validator,
+		Version:   &cparams.Version,
+	}
+
+	// init chain will set the validator set and initialize the genesis accounts
+	testApp.InitChain(
+		abci.RequestInitChain{
+			Time:            gen.GenesisTime,
+			Validators:      []abci.ValidatorUpdate{},
+			ConsensusParams: abciParams,
+			AppStateBytes:   genDoc.AppState,
+			ChainId:         genDoc.ChainID,
+		},
+	)
+
+	// commit genesis changes
+	testApp.Commit()
+	testApp.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{
+		ChainID:            ChainID,
+		Height:             testApp.LastBlockHeight() + 1,
+		AppHash:            testApp.LastCommitID().Hash,
+		ValidatorsHash:     genDoc.ValidatorHash(),
+		NextValidatorsHash: genDoc.ValidatorHash(),
+		Version: tmversion.Consensus{
+			App: cparams.Version.AppVersion,
+		},
+	}})
 
+	return gen.Keyring(), gen.Accounts(), nil
+}
+
+// NewTestAppWithGenesisSet initializes a new app with a validator set and genesis accounts.
```

Comment:
[nit] since this function doesn't take a validator set as a parameter, I think the current godoc is misleading.

```suggestion
// NewTestAppWithGenesisSet initializes a new app with genesis accounts and returns the testApp, validator set and keyring.
```

Reply:
i was not only referring to the arguments it takes but also the overall functionality of the method and what it initialises. happy to change it if you think it's still misleading.

Reply:
I think it's misleading so I prefer the comment I proposed but not blocking on it.